runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -290,45 +267,14 @@ MM_VLHGCAccessBarrier::jniGetPrimitiveArrayCritical(J9VMThread* vmThread, jarray
 		*isCopy = JNI_FALSE;
 	}
 
-	if (alwaysCopyInCritical) {
+	if (alwaysCopyInCritical || !indexableObjectModel->isInlineContiguousArraylet(arrayObject)) {
+		/* alwaysCopyInCritical or discontiguous (including 0 size array) */
 		copyArrayCritical(vmThread, indexableObjectModel, functions, &data, arrayObject, isCopy);
-	} else if (!indexableObjectModel->isInlineContiguousArraylet(arrayObject)) {
-		/* an array having discontiguous extents is another reason to force the critical section to be a copy */
-#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
-		MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(vmThread);
-		if (indexableObjectModel->isDoubleMappingEnabled()) {
-			fj9object_t *arrayoidPtr = indexableObjectModel->getArrayoidPointer(arrayObject);
-			if (indexableObjectModel->isArrayletDataDiscontiguous(arrayObject)) {
-				GC_SlotObject objectSlot(env->getOmrVM(), arrayoidPtr);
-				J9Object *firstLeafSlot = objectSlot.readReferenceFromSlot();
-				MM_HeapRegionDescriptorVLHGC *firstLeafRegionDescriptor = (MM_HeapRegionDescriptorVLHGC *)_extensions->heapRegionManager->tableDescriptorForAddress(firstLeafSlot);
-				data = firstLeafRegionDescriptor->_arrayletDoublemapID.address;
-
-				if (NULL == data) {
-					/* Doublemap failed, but we still need to continue execution; therefore fallback to previous approach */
-					copyArrayCritical(vmThread, indexableObjectModel, functions, &data, arrayObject, isCopy);
-				}
-			/* Corner case where there's only one arraylet leaf */
-			} else if (indexableObjectModel->isArrayletDataContiguous(arrayObject)) {
-				/* Solo arraylet leaf is contiguous so we can simply return the data associated with it */
-				MM_JNICriticalRegion::enterCriticalRegion(vmThread, true);
-				Assert_MM_true(vmThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS);
-				GC_SlotObject objectSlot(env->getOmrVM(), arrayoidPtr);
-				data = objectSlot.readReferenceFromSlot();
-			} else {
-				/* Possible to reach here if arraylet has no leaves.
-				 * Even though there are no elements in it the caller expects a non NULL value
-				 * therefore we just return the address after the object header. */
-				data = (void *)arrayoidPtr;
-				Assert_MM_true((0 == indexableObjectModel->numArraylets(arrayObject)) && (0 == indexableObjectModel->getSizeInElements(arrayObject)));
-			}
-		} else
-#endif /* J9VM_GC_ENABLE_DOUBLE_MAP */
-		{
-			copyArrayCritical(vmThread, indexableObjectModel, functions, &data, arrayObject, isCopy);
-		}
+	} else if (indexableObjectModel->isVirtualLargeObjectHeapEnabled() && !indexableObjectModel->isDataAdjacentToHeader(arrayObject)) {
+		/* off heap enabled and not adjacent */
+		data = (void *)indexableObjectModel->getDataPointerForContiguous(arrayObject);
 	} else {
-		/* acquire access and return a direct pointer */
+		/* adjacent, acquire access and return a direct pointer */
```

Comment:
you can exapand the comment:
- adjcent for offheap enabled and contiguous for offheap disabled
- acquire access to prevent object being moved by GC